src/values/call_site_value.rs

```diff
@@ -210,6 +210,56 @@ impl<'ctx> CallSiteValue<'ctx> {
         }
     }
 
+    /// Get all `Attribute`s on this `CallSiteValue` at an index.
+    ///
+    /// # Example
+    ///
+    /// ```no_run
+    /// use inkwell::attributes::AttributeLoc;
+    /// use inkwell::context::Context;
+    ///
+    /// let context = Context::create();
+    /// let builder = context.create_builder();
+    /// let module = context.create_module("my_mod");
+    /// let void_type = context.void_type();
+    /// let fn_type = void_type.fn_type(&[], false);
+    /// let fn_value = module.add_function("my_fn", fn_type, None);
+    /// let string_attribute = context.create_string_attribute("my_key", "my_val");
+    /// let enum_attribute = context.create_enum_attribute(1, 1);
+    /// let entry_bb = context.append_basic_block(fn_value, "entry");
+    ///
+    /// builder.position_at_end(entry_bb);
+    ///
+    /// let call_site_value = builder.build_call(fn_value, &[], "my_fn");
+    ///
+    /// call_site_value.add_attribute(AttributeLoc::Return, string_attribute);
+    /// call_site_value.add_attribute(AttributeLoc::Return, enum_attribute);
+    ///
+    /// assert_eq!(call_site_value.attributes(AttributeLoc::Return), vec![ string_attribute, enum_attribute ]);
+    /// ```
+    #[llvm_versions(3.9..=latest)]
+    pub fn attributes(self, loc: AttributeLoc) -> Vec<Attribute> {
+        use llvm_sys::core::LLVMGetCallSiteAttributes;
+        use std::mem::MaybeUninit;
+
+        let count = self.count_attributes(loc) as usize;
+
+        // initialize a vector, but leave each element uninitialized
+        let mut attribute_refs: Vec<MaybeUninit<Attribute>> = vec![MaybeUninit::uninit(); count];
+
+        // Safety: relies on `Attribute` having the same in-memory representation as `LLVMAttributeRef`
+        unsafe {
+            LLVMGetCallSiteAttributes(
+                self.as_value_ref(),
+                loc.get_index(),
+                attribute_refs.as_mut_ptr() as *mut _,
+            )
+        }
+
+        // Safety: everything is initialized
+        unsafe { std::mem::transmute(attribute_refs) }
```

Comment:
Conceptually I'm fine with this change, but we should avoid transmute at all costs imo. The transmute docs do have an example on converting a `Vec<A>` to `Vec<B>` without transmute that we should try and leverage: https://doc.rust-lang.org/std/mem/fn.transmute.html

(search for "This is the proper no-copy, unsafe way of "transmuting" a `Vec`")